### PR TITLE
Feature/batched calculations

### PR DESCRIPTION
**Pull request by benblack769**
I noticed that PPO was using far more memory than I was expecting, and tracked it down to a number of no_grad and target operations in the generalized advantage wrapper and the PPO agent that had no batching, i.e. all the data was being processed at once, leading to huge blowups in memory, even when `minibatches` was large.

This PR has my proposed solution. It extends the StateArray object with 3 new methods:

* cat (class method): concatenates StateArrays
* state_array.execute_batched(batch_size, fn): executes function batch-wise on data, returns the concatenated state array
* state_array.flatten_for_execution(fn): flattens data before performing function. Used in the generalized advantage buffer so that execute_batched can be successfully applied.

### CODE REVIEW

**Comment by cpnota** [comment]
Interesting... will need to look at this more later today. I like that the solution is reusable, but need to to think about the changes to the interface a little bit.

File: all/core/state.py [hunk 1]
```diff
@@ -387,6 +387,49 @@ def shape(self):
...
+        state_size = sum(state_array.shape[axis] for state_array in state_array_list)
+        new_shape = list(state_array_list[0].shape)
+        new_shape[axis] = state_size
+        new_shape = tuple(new_shape)
+        keys = list(state_array_list[0].keys())
+        for key in keys:
+            d[key] = torch.cat([state_array[key] for state_array in state_array_list], axis=axis)
+        return StateArray(d, new_shape, device=state_array_list[0].device)
+
+    def flatten_for_execution(self, fn):
```
Comment (cpnota) [inline all/core/state.py:406]: This function seems a bit contrived.  Would it be clearer to just add a `flatten` method and let users handle resizing for `batch_execute`?  

```
target_values = states.flatten().batch_execute(256, compute_target_values).view(states.shape)
```

File: all/core/state.py [hunk 2]
```diff
@@ -387,6 +387,49 @@ def shape(self):
...
+        flat_size = np.prod(self.shape)
+        flat_data = self.view((flat_size,))
+        out = fn(flat_data)
+        assert out.shape[0] == flat_size, "first dimension out output of flatten_for_execution expected to match input"
+        spread = out.view(self.shape + out.shape[1:])
+        return spread
+
+    def batch_execute(self, minibatch_size, fn):
```
Comment (cpnota) [inline all/core/state.py:414]: I'm wondering if `minibatch_size` should optionally accept `-1` to indicate no mini-batching, or something to that effect. Or I think it can handle infinity? Is that usage clear?

File: all/experiments/multiagent_env_experiment.py [hunk 3]
```diff
@@ -29,7 +29,7 @@ def __init__(
             name=None,
             quiet=False,
             render=False,
-            save_freq=100,
+            save_freq=float('inf'),
```
Comment (cpnota) [inline all/experiments/multiagent_env_experiment.py:32]: This looks left over from the other PR. Do you have strong opinions about merging that?

**Review by cpnota — COMMENTED** [review]
Thanks as always for a thoughtful PR! I like the approach. I just have a few questions.

File: all/core/state.py (hunk 1 above)
Comment (benblack769) [reply all/core/state.py:406]: Yeah, that is probably a better approach.

File: all/core/state.py (hunk 2 above)
Comment (benblack769) [reply all/core/state.py:414]: I think you can put in an arbitrarily large number. Probably a float infinity would work. So I think it is fine as is?

File: all/experiments/multiagent_env_experiment.py (hunk 3 above)
Comment (benblack769) [reply all/experiments/multiagent_env_experiment.py:32]: No, I just branched off of my old branch instead of develop. Whoops. I can change this back.

File: all/core/state.py [hunk 4]
```diff
@@ -387,6 +387,49 @@ def shape(self):
     def __len__(self):
         return self.shape[0]
 
+    @classmethod
+    def cat(cls, state_array_list, axis=0):
```
[the remark this answers is not in the document]
Comment (benblack769) [reply all/core/state.py:391]: @cpnota I would comment that I think that tianshau's Batch https://tianshou.readthedocs.io/en/master/tutorials/batch.html is a more general, easier to use, and more sensible data structure than ALL's state objects, that already have concatenation and stacking functionality built in. I think that ALL would benefit significantly by either adopting tianshau's Batch structure or rethinking states to effectively do the same thing.

File: all/experiments/multiagent_env_experiment.py (hunk 3 above)
Comment (benblack769) [reply all/experiments/multiagent_env_experiment.py]: Reverted

File: all/core/state.py (hunk 1 above)
Comment (benblack769) [reply all/core/state.py]: Some weird things happening here in the unit tests I can't quite figure out.

File: all/core/state.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cpnota) [reply all/core/state.py:391]: Improving `State` is definitely something on my backlog, thanks for the reference!

**Review by cpnota — APPROVED** [review]
